pmd-core/src/main/java/net/sourceforge/pmd/cache/AbstractAnalysisCache.java

```diff
@@ -153,28 +154,48 @@ public void checkValidity(final RuleSets ruleSets, final ClassLoader auxclassPat
         ruleMapper.initialize(ruleSets);
     }
 
+    private static boolean isClassPathWildcard(String entry) {
+        return entry.endsWith("/*") || entry.endsWith("\\*");
+    }
+
     private URL[] getClassPathEntries() {
         final String classpath = System.getProperty("java.class.path");
         final String[] classpathEntries = classpath.split(File.pathSeparator);
         final List<URL> entries = new ArrayList<>();
 
+        final SimpleFileVisitor<Path> fileVisitor = new SimpleFileVisitor<Path>() {
+            @Override
+            public FileVisitResult visitFile(final Path file,
+                    final BasicFileAttributes attrs) throws IOException {
+                if (!attrs.isSymbolicLink()) { // Broken link that can't be followed
+                    entries.add(file.toUri().toURL());
+                }
+                return FileVisitResult.CONTINUE;
+            }
+        };
+        final SimpleFileVisitor<Path> jarFileVisitor = new SimpleFileVisitor<Path>() {
+            @Override
+            public FileVisitResult visitFile(final Path file,
+                    final BasicFileAttributes attrs) throws IOException {
+                String extension = FilenameUtils.getExtension(file.toString());
+                if ("jar".equalsIgnoreCase(extension)) {
```

Comment:
we could also find raw `.class` files here, and we should visit those too... `.zip` is also valid as per https://docs.oracle.com/javase/7/docs/technotes/tools/windows/classpath.html

Reply:
nevermind, that same doc says:

> Class path entries can contain the basename wildcard character *, which is considered equivalent to specifying a list of all the files in the directory with the extension .jar or .JAR

> A class path entry that contains * will not match class files.